### ISSUE TITLE
Chunks and assets iteration over IPFS & Never ending Process

### ISSUE THREAD

**Issue by Asone** [VictorBjelkholm/ipfs-webpack-plugin#1]
Hi, 

I made a few experimentations in order to fix two bugs annoted in your version of the plug-in. 

I had to rewrite most of the plug-in in order to be able to solve the bugs and this version doesn't do anything specific with the IFPS results. 

I don't know if that could fit to your needs but i thought that might help. You can see the [modified version here](https://github.com/Asone/ipfs-webpack-plugin/blob/bugfix/file_iteration/ipfs-webpack-plugin.js)

####  Never ending process 

Bug has been fixed by injecting an instance of IPFS into plugin's instance without start. 

` IPFSWebpackPlugin.prototype._ipfs = new IPFS({start: false});`

Once the compilation done, the IPFS instance is started and assets are iterated for processing in promises. 

```
this._ipfs.start() // start ipfs node
        .then(res => {})
        .catch(res => {});
```

The promises chain allows us to stop the IPFS instance once our assets are processed over IPFS.

```
this._ipfs.start() // start ipfs node
        .then(res => {})
        .then( res => {
             this._ipfs.stop();
        })
        .catch(res => {});
```

####  assets iteration

Bug has been fixed by storing the list of assets generated into a plugin scoped variable.
Once compilation done, we iterate through each asset in order to add it through the ipfs instance. 

The promises chain will allow to process the result of the added file to IPFS. Next step here is to inject the chunk assets into the index file generated through webpack. 

In any case, thanks for the original development ! 👍

**Comment by Asone** [comment]
After a few tests, i realize that my version does not take in consideration the assets loaded in params ( like the ipfs.js and loader.js ) but only the chunks of the declared entrypoints. 

`params.assets` has a key for every assets with the name of the file as key. I'll try to find a way to iterate over those by getting the key name & rebuild the paths. 

Event though, i'm not sure the plug-in itself should rely on loading itself some assets, but rather declare those in the webpack config as chunks.

**Comment by Asone** [comment]
I updated the code on my side. 

This time i get the keys of the assets object provided by params  as `keys` are the filenames and iterate over those, taking care in the iteration to remove the appended parameters on the assets names. 

```
const keys = Object.keys(this.params.assets);

        for(asset of keys){ // iterate through assets
          const filepath = asset.replace(/\?.*$/,''); // clean params in filename
          fileToReplaceIn = (outputOptions.path + '/' + filepath); // build environmental path
          const file = fs.readFileSync(fileToReplaceIn); // loads file
          
          this._ipfs.files.add({ path: filepath, content: file })
        }
```

EDIT : I'll still need to do some tests on more complex configuration with multiple entrypoints, multiple filetypes, in-depth dist assets output and chunks splitting. 

EDIT 2 : ~Also, i haven't tested yet with a watcher on the webpack runner~
Tested with watch option, seems to run fine.

**Comment by Asone** [comment]
I just tried adding files to the repo with the plugin but without starting the IPFS node. It seems we effectively get the hashes and files gets added to the ipfs-repo. 

@VictorBjelkholm : At this point, i'm asking myself if another IPFS instance, outside of the plugin, but connected, would automatically detect the added files to the repo if both of the instances are pointing to the same repo ? 

I'm asking because in that case, imho it would make more sense then to have a plugin that does not effectively connect to the network but just adds the files. 

Problem is the following :
 If we want the webpack instance to close, we need to have the webpack plugin instance to close its instance of IPFS, which means no assets available on the network + the latency beetween starting node and stopping it can be very short if there's very few assets, for which i'm not sure it's interesting. 

Expecting user to have its own parallel instance and just adding the files to its repo would allow him/her to have a synced node that would allow access of the assets generated and added by the plugin. 

Also, i've began to check about the HTMLWebpackPlugin addition, and will try to release some attempts of implementation in the next weeks.

**Comment by crazyrabbitLTC** [comment]
Hello Asone, any update? I'm curious about this idea.

**Comment by Asone** [comment]
Hi @crazyrabbitLTC,
Thanks for asking !

unfortunately there hasn't been valuable progress since then. 

TL;DR this repo seems to be an abandoned PoC for now and there's nothing i can really use right now. original intent from @VictorBjelkholm adds static assets to repo a local repo and should be functional. 

My intention was to publish the generated assets on IPFS from a compilation and replace main uris pointers to the returned IPFS locations.

I managed to somewhat fix those bugs but lacked the necessary knowledge to  produce a satisfiable code as there are many complexity points.

I've explored different options in order to research and test which i could document if you want and concluded that achieving an advanced orchestration beetween webpack and IPFS appears to be quite challenging.

At this point i'm not sure it's the more affordable way to deploy a front-end to IPFS network, and would first check out for alternative processes.

**Comment by Asone** [comment]
@crazyrabbitLTC : you might want to give a try to [embark](https://github.com/embark-framework/embark) as it seems it is built to allow easy deployment for both back-end ( eth ) and front-end ( ipfs ) of Dapps.
